fix(docs): improve image docs from workflow evaluation (#14697)
gatsbyjs/gatsby: docs/docs/gatsby-image-api.md

---

> title: Gatsby Image API

Reviewer: I like the title of "Gatsby Image API", but I think the URL should be gatsbyjs.org/docs/gatsby-image — maybe rename this to `gatsby-image.md`?

Author: Sure, we can do that. I picked that URL to differentiate from other files that already exist, but it isn't consistent with the others.

---

Part of what makes Gatsby sites so fast is its recommended approach to handling images. `gatsby-image` is a React component designed to work seamlessly with Gatsby’s [native image processing](https://image-processing.gatsbyjs.org/) capabilities powered by GraphQL and [gatsby-plugin-sharp](/packages/gatsby-plugin-sharp/) to easily and completely optimize image loading for your sites.

> > _Warning: gatsby-image is **not** a drop-in replacement for `<img />`. It’s optimized for fixed width/height images and images that stretch the full-width of a container._

Reviewer: "Warning" makes it sound like it might be scary — swapping it out for a note makes it more of a heads-up.
```suggestion
> _Note: gatsby-image is **not** a drop-in replacement for `<img />`. It’s optimized for fixed width/height images and images that stretch the full-width of a container._
```

Demo: [https://using-gatsby-image.gatsbyjs.org/](https://using-gatsby-image.gatsbyjs.org/)

## In this doc

- [Setting up Gatsby Image](#setting-up-gatsby-image)

> - [Types of gatsby-images](#types-of-gatsby-images)

Reviewer: ```suggestion
- [Types of Gatsby images](#types-of-gatsby-images)
```

  - [Fixed images and parameters](#images-with-a-fixed-width-and-height)
  - [Fluid images and parameters](#images-that-stretch-across-a-fluid-container)
  - [Resized images](#resized-images)
  - [Shared query parameters](#shared-query-parameters)
- [Image query fragments](#image-query-fragments)
- [Gatsby Image props](#gatsby-image-props)

## Setting up Gatsby Image

> To install Gatsby Image, install the `gatsby-image` package along with necessary plugins `gatsby-transformer-sharp` and `gatsby-plugin-sharp`. Reference the packages in your Gatsby config file. You can also provide additional options to [`gatsby-plugin-sharp`](/packages/gatsby-plugin-sharp/) in your config file.

Reviewer: Should we include a copy-pasteable block here for convenience?

Author: I debated on that. It's included so many places, I linked to the docs instead. But we could add it if necessary

Author: I ended up adding it in! Thanks for the nudge.

A common way to source images is to install and use `gatsby-source-filesystem` to connect your local files, but other source plugins can be used as well, such as `gatsby-source-contentful`, `gatsby-source-datocms` and `gatsby-source-sanity`.

_For in-depth install instructions, check out the docs on [Using Gatsby Image](/docs/using-gatsby-image/)._

### Gatsby image starts with a query

To feed image data in to Gatsby Image, set up a [GraphQL query](/docs/graphql-reference/) and either pass it into a component as props or write it in the component itself. One technique is to leverage the [`useStaticQuery`](/docs/use-static-query/) hook.

Common GraphQL query methods for sourcing images include `file` from [gatsby-source-filesystem](/packages/gatsby-source-filesystem/), and both `imageSharp` and `allImageSharp` from [gatsby-plugin-sharp](/packages/gatsby-plugin-sharp/), but ultimately the API methods available to you will depend on your content source.

> You can also use [GraphQL aliases](/docs/graphql-reference/#aliasing) for querying multiple images of the same type.

Reviewer: Should this be a note/aside?

Author: Sure, good idea.

See below for code examples of queries and how to use them in components.

> ## Types of gatsby-images

Reviewer: ```suggestion
## Types of Gatsby images
```

> Gatsby-image objects are created through GraphQL methods. There are two types of responsive images available, _fixed_ and _fluid_, which create multiple image sizes (1x, 1.5x, etc.). There is also the _resize_ method, which returns a single image.

Reviewer: ```suggestion
Gatsby image objects are created through GraphQL methods. There are two types of image optimizations available: _fixed_ and _fluid_, which create multiple image sizes (1x, 1.5x, etc.). There is also the _resize_ method, which returns a single image.
```

import Img from "gatsby-image"

export default () => {
  const data = useStaticQuery(query`
    file(relativePath: { eq: "images/default.jpg" }) {
      childImageSharp {
        # Specify a fixed image and fragment
        // highlight-start
        fixed {
          ...GatsbyImageSharpFixed
        }
        // highlight-end
      }
    }
  `)
  return (
    <div>
      <h1>Hello gatsby-image</h1>
      <Img
        fixed={data.file.childImageSharp.fixed} {/* highlight-line */}
        alt="Gatsby Docs are awesome"
      />
    </div>
  )
}
```

Read more in the [gatsby-plugin-sharp](/packages/gatsby-plugin-sharp/?=#fixed) README.

#### Fixed image query parameters

In a query, you can specify options for fixed images.

- `width` (int, default: 400)
- `height` (int)
- `quality` (int, default: 50)

#### Returns

- `base64` (string)
- `aspectRatio` (float)
- `width` (float)
- `height` (float)
- `src` (string)
- `srcSet` (string)

This is where fragments like `GatsbyImageSharpFixed` come in handy, as they'll return all the above items in one line:

```graphql
file(relativePath: { eq: "images/default.jpg" }) {
  childImageSharp {
    # Specify a fixed image and fragment

>     // highlight-next-line

Reviewer: ```suggestion
    // highlight-start
```

    fixed(width: 400, height: 400) {
      ...GatsbyImageSharpFixed

>     }

Reviewer: ```suggestion
    }
    // highlight-end
```

  }
}
```

### Images that stretch across a _fluid_ container

Create fluid sizes (in width) for the image. E.g. if the max width of the container for a rendered Markdown file is 800px, the automatic sizes would then be: 200px, 400px, 800px, 1200px, 1600px – enough to provide close to the optimal image size for every device size / screen resolution. If you want more control over which sizes are output you can use the `srcSetBreakpoints` parameter.

Once you've queried for an image to retrieve its data, you can pass the `fluid` data into the `Img` component:

```jsx
import { useStaticQuery } from "gatsby"
import Img from "gatsby-image"

export default () => {
  const data = useStaticQuery(query`
    file(relativePath: { eq: "images/default.jpg" }) {
      childImageSharp {
        # Specify a fluid image and fragment
        // highlight-start
        fluid {
          ...GatsbyImageSharpFluid
        }
        // highlight-end
      }
    }
  `)
  return (
    <div>
      <h1>Hello gatsby-image</h1>
      <Img
        fluid={data.file.childImageSharp.fluid} {/* highlight-line */}
        alt="Gatsby Docs are awesome"
      />
    </div>
  )
}
```

> Read more in the [gatsby-plugin-sharp](/packages/gatsby-plugin-sharp/?=#fluid) README.

Reviewer: ```suggestion
Read more in the [gatsby-plugin-sharp](/packages/gatsby-plugin-sharp/#fluid) README.
```

Reviewer: Anchors don't seem to be working, FYI. 😕

Author: They don't seem to work for me no matter what the URL is...

Author: Actually they do work for me in their existing form, at least locally when I click them. They don't work when deep-linked

#### Fluid image query parameters

In a query, you can specify options for fluid images.

- `maxWidth` (int, default: 800)
- `maxHeight`(int)
- `quality` (int, default: 50)
- `srcSetBreakpoints` (array of int, default: [])
- `fit` (string, default: `[sharp.fit.cover][6]`)
- `background` (string, default: `rgba(0,0,0,1)`)

#### Returns

- `src` (string)
- `width` (int)
- `height` (int)

> - `aspectRatio` (float)

Reviewer: This list is missing `base64`, `srcSet`, etc.

Author: nice catch!

file(relativePath: { eq: "images/default.jpg" }) {
  childImageSharp {
    // highlight-next-line
    fluid(maxWidth: 400) {
      // highlight-next-line
      ...GatsbyImageSharpFluid
    }
  }
}
```

### Resized images

In addition to _fixed_ and _fluid_ images, the gatsby-image API allows you to call a `resize` method with `gatsby-plugin-sharp` to return a single image as opposed to multiple sizes. There are no default fragments available for the resize method.

#### Parameters

- `width` (int, default: 400)
- `height` (int)
- `quality` (int, default: 50)
- `jpegProgressive` (bool, default: true)
- `pngCompressionLevel` (int, default: 9)
- `base64`(bool, default: false)

#### Returns

Resize returns an object with the following items:

- `src` (string)
- `width` (int)
- `height` (int)
- `aspectRatio` (float)

```graphql
allImageSharp {
  edges {
    node {
        resize(width: 150, height: 150, grayscale: true) {
          src
        }
    }
  }
}
```

### Shared query parameters

In addition to `gatsby-plugin-sharp` settings in `gatsby-config.js`, there are additional query options that apply to both _fluid_ and _fixed_ images:

- `grayscale` (bool, default: false)

> - `duotone` (bool|obj, default: false)

Reviewer: This might be too much, but a before/after of `grayscale` and `duotone` would be dope.

Author: ooh maybe @gillkyle could help me with that!

Reviewer 2: Here's what I added (note: these images will only be displayed like this if you've upgraded `gatsby-remark-images` like this PR #14716):
![example-images-screenshot](https://user-images.githubusercontent.com/21114044/59311683-a8060e00-8c67-11e9-9aee-5ccdf970cb57.png)

Just pushed right to this branch, feel free to edit it if you want.

Author: you're amazing, thanks @gillkyle !

## Gatsby-image props

After you've made a query, you can pass additional options to the gatsby-image component.

| Name                   | Type                | Description                                                                                                                   |
| ---------------------- | ------------------- | ----------------------------------------------------------------------------------------------------------------------------- |
| `fixed`                | `object`            | Data returned from the `fixed` query                                                                                          |
| `fluid`                | `object`            | Data returned from the `fluid` query                                                                                          |
| `fadeIn`               | `bool`              | Defaults to fading in the image on load                                                                                       |

> | `durationFadeIn`       | `number`            | fading duration is set up to 500ms by default                                                                                 |

Reviewer: ```suggestion
| `durationFadeIn`       | `number`            | fade-in duration is set to 500ms by default                                                                                 |
```
